dlmalloc/src/malloc.c

```diff
@@ -4588,6 +4593,13 @@ void* dlmalloc(size_t bytes) {
   ensure_initialization(); /* initialize in sys_alloc if not using locks */
 #endif
 
+#if __wasilibc_unmodified_upstream // Try to initialize the allocator.
+#else
+  if (!is_initialized(gm)) {
+    try_init_allocator();
+  }
+#endif
```

Comment:
Would it work to call this from within `init_mparams` instead? That way it'd get called for all entrypoints, not just `malloc`.

In that case, it shouldn't do `ensure_initialization` itself, but just do its work after `init_mparams` has done its work.

Reply:
Hmmm, i've though about it, but there are some clues:
1. `ensure_initialization` is used for mspace (mspace is like Arena in malloc for different continuous space of allocations). At now, mspace is disabled, but if someone want to enable it, there will some unobvious problems. And it is logically incorrect.
2. Currently, `ensure_initialization` is disabled in `dlmalloc`, because it is inside `#if USE_LOCKS` that is also disabled. And it is called into `sys_alloc`. But for our purpose it is important to have it initialized before main logic of chunk choosing in `dlmalloc` being called. => in case of  `ensure_initialization` we will need two augmentation (into `ensure_initialization` and `dlmalloc`) instead of one in current case.
3. On the other side this initialization into `dlmalloc` seems correct, because other entry point (`realloc`, `calloc`) in all important CFG paths will call it before any chunk manipulation. The only exception here is `free`, but it is UB to call `free` without any previous `malloc`.

So, yes, it possible to move top chunk initialization into `insure_initialization`, but it seems that it is better leave it as now.

Reply:
Thanks for looking into that! I think what you have right now looks like a reasonable approach then.